File: apps/windows/TeslaSync.App/feature-views/ResetSection/ResetSection.Source.cs
using System.Text.Json;
using TeslaSync.App.Core.Data.Net;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutation port the <see cref="ResetSectionViewModel"/> binds to (P1/S8 state-holder seam) — the native
/// analogue of the web <c>useResetSection</c> / <c>useResetAllSettings</c> hooks
/// (web/src/api/hooks/useSettingsReset.ts). Both drive the same <c>POST /settings/reset</c> endpoint: a
/// single-section reset sends <c>{ section }</c>, the danger-zone reset sends <c>{}</c>. The view never
/// performs HTTP itself; the concrete <see cref="SettingsResetSource"/> (or a test fake) drives this.
/// </summary>
public interface ISettingsResetSource
{
    /// <summary>
    /// Reset one whitelisted section (web <c>useResetSection(section).mutate()</c>): <c>POST /settings/reset</c>
    /// with <c>{ section }</c>. Throws on a transport / HTTP failure so the caller can surface the error.
    /// </summary>
    Task<SettingsResetResult> ResetSectionAsync(string section, CancellationToken cancellationToken = default);

    /// <summary>
    /// Reset every whitelisted section (web <c>useResetAllSettings().mutate()</c>): <c>POST /settings/reset</c>
    /// with an empty body. Throws on a transport / HTTP failure so the caller can surface the error.
    /// </summary>
    Task<SettingsResetResult> ResetAllAsync(CancellationToken cancellationToken = default);
}

/// <summary>
/// The contract-client-backed <see cref="ISettingsResetSource"/> — the native data adapter for the reset
/// surface. Both operations POST the generated <c>post_api_v1_settings_reset</c> endpoint through the shared
/// <see cref="IApiClient"/> (the same auth + resilience pipeline the rest of the app shares; the web's
/// transparent SUDO step-up is handled by the shared auth handler) and parse the JSON receipt via
/// <see cref="SettingsResetResultParser"/>. The body is a plain string map so the wire shape matches the web
/// exactly — <c>{"section":"…"}</c> for a section reset and <c>{}</c> for the global reset — regardless of the
/// shared serializer's naming policy. No HTTP touches the view.
/// </summary>
public sealed class SettingsResetSource : ISettingsResetSource
{
    /// <summary>The generated OpenAPI operation id for <c>POST /api/v1/settings/reset</c>.</summary>
    private const string ResetOperation = "post_api_v1_settings_reset";

    private readonly IApiClient _api;

    /// <summary>Creates the source over the shared contract client.</summary>
    /// <param name="api">The generated contract client used for the reset POST.</param>
    public SettingsResetSource(IApiClient api)
    {
        ArgumentNullException.ThrowIfNull(api);
        _api = api;
    }

    /// <inheritdoc />
    public Task<SettingsResetResult> ResetSectionAsync(string section, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(section);
        var body = new Dictionary<string, string>(StringComparer.Ordinal) { ["section"] = section };
        return SendAsync(body, cancellationToken);
    }

    /// <inheritdoc />
    public Task<SettingsResetResult> ResetAllAsync(CancellationToken cancellationToken = default) =>
        SendAsync(new Dictionary<string, string>(StringComparer.Ordinal), cancellationToken);

    private async Task<SettingsResetResult> SendAsync(
        IReadOnlyDictionary<string, string> body,
        CancellationToken cancellationToken)
    {
        var request = new ApiRequest(ResetOperation, Body: body);
        var response = await _api.SendAsync<JsonElement>(request, cancellationToken).ConfigureAwait(false);
        return SettingsResetResultParser.Parse(response);
    }
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/ResetSection/ResetSectionSource.kt
// The data port the Reset-to-defaults surface binds to (P1/S8 state-holder seam) — the native analogue of the
// two web hooks the panel owns: `useResetSection` and `useResetAllSettings`
// (web/src/features/settings/components/ResetSection.tsx, web/src/api/hooks/useSettingsReset.ts). The view
// never performs HTTP (ADR-002); a concrete adapter over the shared S8
// [io.teslasync.shared.core.presentation.settingsreset.SettingsResetStore] (or a test fake) drives this seam.
//
// Both web hooks are `useMutation`s with NO `useQuery`, so the seam is two non-throwing suspend [Result]
// mutations and carries no cache-then-network feed. On success the shared repository has already flushed the
// whole offline cache (the data-layer analogue of the web hooks' argument-less `queryClient.invalidateQueries()`)
// and the shared store has cached the receipt — neither is this surface's concern. The backend path is
// sudo-gated, but the shared `ApiHttpClient` handles the `RequireSudo` step-up transparently via its sudo-token
// sink, so — unlike the web `request()` client's interactive ReauthDialog — there is no distinct "sudo
// canceled" outcome here: a failed step-up simply surfaces as a failed [Result], which the surface toasts.
//
// `InvalidPackageDeclaration` is suppressed: the mandated surface directory
// (com/teslasync/feature-views/ResetSection) cannot form a valid Kotlin package, so the package intentionally
// diverges from the path. `MatchingDeclarationName` is suppressed: the mandated `ResetSection*` filename
// cannot match the seam's `ResetSectionSource` name.
@file:Suppress("InvalidPackageDeclaration", "MatchingDeclarationName")

package io.teslasync.android.featureviews.resetsection

import io.teslasync.shared.core.presentation.settingsreset.SettingsResetResult
import io.teslasync.shared.core.presentation.settingsreset.SettingsResetSectionResult
import io.teslasync.shared.core.presentation.settingsreset.SettingsResetStore

/**
 * The settings-reset mutation seam the surface binds to — the native port of the web `useResetSection` +
 * `useResetAllSettings` hooks. A narrow two-method port so the view-model depends on an abstraction (the
 * shared-layer adapter in production ↔ a fake in tests), never on the store or the HTTP client directly. Both
 * methods are `suspend` non-throwing [Result]s mirroring the web hooks' `mutateAsync`.
 */
interface ResetSectionSource {
    /**
     * Resets one named section (web `useResetSection` → `POST /settings/reset { section }`). [section] is the
     * canonical lower-snake-case name; an unknown/denied section surfaces as a failed [Result].
     */
    suspend fun resetSection(section: String): Result<SettingsResetResult>

    /** Resets every whitelisted section (web `useResetAllSettings` → `POST /settings/reset {}`). */
    suspend fun resetAll(): Result<SettingsResetResult>
}

/**
 * Binds the surface to the shared **S8** [SettingsResetStore] — the single cross-platform holder every reset
 * surface shares (it routes both mutations through the S7 repository, caches the receipt into `lastReset`, and
 * has already flushed the offline cache on success). The production composition wires this; the view receives
 * the resulting [ResetSectionSource].
 */
fun SettingsResetStore.asResetSectionSource(): ResetSectionSource {
    val store = this
    return object : ResetSectionSource {
        override suspend fun resetSection(section: String): Result<SettingsResetResult> = store.resetSection(section)

        override suspend fun resetAll(): Result<SettingsResetResult> = store.resetAll()
    }
}

/**
 * An in-memory [ResetSectionSource] for previews and tests — it records the calls it received and returns the
 * configured [Result]s ([sectionOutcome] / [allOutcome]). Not thread-safe by design (single-writer, like the
 * web component itself). The defaults are a small successful receipt so a preview's confirm flow resolves.
 *
 * @property sectionOutcome the result returned for a [resetSection] call (keyed by the section wire name).
 * @property allOutcome the result returned for a [resetAll] call.
 */
class InMemoryResetSectionSource(
    private val sectionOutcome: (String) -> Result<SettingsResetResult> = { Result.success(DEFAULT_SECTION_RESULT) },
    private val allOutcome: () -> Result<SettingsResetResult> = { Result.success(DEFAULT_ALL_RESULT) },
) : ResetSectionSource {
    private val recordedSectionCalls = mutableListOf<String>()
    private var recordedAllCalls = 0

    /** The section wire names passed to [resetSection], in call order (test assertion seam). */
    val sectionCalls: List<String> get() = recordedSectionCalls.toList()

    /** The number of [resetAll] calls received (test assertion seam). */
    val allCalls: Int get() = recordedAllCalls

    override suspend fun resetSection(section: String): Result<SettingsResetResult> {
        recordedSectionCalls += section
        return sectionOutcome(section)
    }

    override suspend fun resetAll(): Result<SettingsResetResult> {
        recordedAllCalls += 1
        return allOutcome()
    }

    private companion object {
        /** A single-section receipt: one section cleared a couple of rows. */
        val DEFAULT_SECTION_RESULT: SettingsResetResult =
            SettingsResetResult(reset = 2, sections = listOf(SettingsResetSectionResult(section = "general", reset = 2)))

        /** A global receipt: a few sections cleared several rows. */
        val DEFAULT_ALL_RESULT: SettingsResetResult =
            SettingsResetResult(
                reset = 7,
                sections =
                    listOf(
                        SettingsResetSectionResult(section = "alert_rules", reset = 3),
                        SettingsResetSectionResult(section = "geofences", reset = 2),
                        SettingsResetSectionResult(section = "automations", reset = 2),
                    ),
            )
    }
}
